_source/_posts/2023-06-22-net-maui-authentication.md

```diff
@@ -0,0 +1,582 @@
+                        }
+
+                    }, cancellationToken).ConfigureAwait(false);
+
+                if (discoveryDocumentResponse.IsError)
+                {
+                    throw new InvalidOperationException("Error loading discovery document: " +
+                                                        discoveryDocumentResponse.Error, discoveryDocumentResponse.Exception);
+                }
+
+                _oidcClient.Options.ProviderInformation = new ProviderInformation
+                {
+                    IssuerName = discoveryDocumentResponse.Issuer,
+                    KeySet = discoveryDocumentResponse.KeySet,
+                    AuthorizeEndpoint = discoveryDocumentResponse.AuthorizeEndpoint,
+                    TokenEndpoint = discoveryDocumentResponse.TokenEndpoint,
+                    EndSessionEndpoint = discoveryDocumentResponse.EndSessionEndpoint,
+                    UserInfoEndpoint = discoveryDocumentResponse.UserInfoEndpoint,
+                    TokenEndPointAuthenticationMethods =
+                        discoveryDocumentResponse.TokenEndpointAuthenticationMethodsSupported,
+
+                };
+            }
+        }
+    }
+}
+```
+
+Before jumping into the `OktaClientConfiguration` creation, I'd like to explain a few differences between this implementation and the Auth0 client described in [Add Authentication to .NET MAUI Apps with Auth0](https://auth0.com/blog/add-authentication-to-dotnet-maui-apps-with-auth0/) blog post from Auth0 by Okta.
+
+Okta provides [two different types of authorization servers](https://developer.okta.com/docs/concepts/auth-servers/), the "Org Authorization Server" (Org AS) and the "Custom Authorization Server" (Custom AS). Suppose you try to use a Custom AS with IdentityModel. In that case. In that case, you will face the following error "[Error loading discovery document: Endpoint belongs to different authority: https://xxxxxxxxx.com/oauth2/v1/clients](https://devforum.okta.com/t/error-loading-discovery-document-endpoint-belongs-to-different-authority/6582)". This is because, during discovery, IdentityModel validates that all endpoints belong to the same authorization server, and with Okta, the client registration occurs at the Org level, not the authorization server level.
+
+In Okta, applications are global to the Org and can be used for multiple authorization servers; that's why you cannot register a client for only one Authorization Server and, ultimately, why the path in metadata is not specific to the Authorization Server that the discovery request was for.
+To avoid the above mentioned issue, we must implement our discovery logic and tell IdentityModel that endpoints can belong to two different authorization servers. The method `EnsureProviderInformationAsync` contains the custom discovery logic, and we need to make sure to call this method before the actual Login and Logout.
```

Comment:
```suggestion
To avoid the issue mentioned above, we must implement our discovery logic and tell IdentityModel that endpoints can belong to two different authorization servers. The method `EnsureProviderInformationAsync` contains the custom discovery logic, and we need to make sure to call this method before the actual Login and Logout.
```